### ISSUE TITLE
Rule proposition: `no-shared-component-data` must be a function.

### ISSUE THREAD

**Issue by armano2** [vuejs/eslint-plugin-vue#79]
When using the data property on a component (i.e. anywhere except on new Vue), the value must be a function that returns an object.

based on [style guide](https://vue-style-guide-dev.surge.sh/v2/guide/style-guide.html#Component-data)

Invalid:
```js
Vue.component('some-comp', {
  data: {
    foo: 'bar'
  }
})
```
```js
export default {
  data: {
    foo: 'bar'
  }
}
```

Valid:
```js
new Vue({
  data: function () {
    return {
      foo: 'bar'
    }
  }
})
```
```js
new Vue({
  data: {
    foo: 'bar'
  }
})
```
```js
Vue.component('some-comp', {
  data: function () {
    return {
      foo: 'bar'
    }
  }
})
```
```js
export default {
  data: function () {
    return {
      foo: 'bar'
    }
  }
}
```

**Comment by armano2** [comment]
Can I start working on this rule?

**Comment by michalsnik** [comment]
I definitely agree with this one 👍 It's something strongly advised in vue documentation and I think it's a must have rule in our set.

**Comment by armano2** [comment]
it's already implemented and `ready for review` 🍡

**Comment by michalsnik** [comment]
Released in `v3.8.0`. Thanks @armano2 !
